src/main/java/it/gov/pagopa/pu/classification/service/BalanceAmountService.java

```diff
@@ -0,0 +1,82 @@
+
+  public BalanceAmountService(BalanceService balanceService, BalanceDefaultMarshallingService balanceDefaultMarshallingService) {
+    this.balanceService = balanceService;
+    this.balanceDefaultMarshallingService = balanceDefaultMarshallingService;
+  }
+
+  public String calculateAmountBalance(CalculateAmountBalanceRequest calculateAmountBalanceRequest) {
+    String balance = calculateAmountBalanceRequest.getBalance();
+    BigDecimal amountInstallment = Utilities.longCentsToBigDecimalEuro(calculateAmountBalanceRequest.getAmountCents());
+
+    Object balanceXML = balanceService.unmarshalBalance(balance);
+
+    if (balanceXML instanceof CtBilancio) {
+      log.info("The balance amount is already calculated");
+      return balance;
+    }
+
+    try {
+      CtBilancioDefault ctBilancioDefault = (CtBilancioDefault) balanceXML;
+      log.info("Calculating balance amount resolving default type");
+      for (CtCapitoloDefault capitolo : ctBilancioDefault.getCapitolo()) {
+        for (CtAccertamentoDefault ctAccertamentoDefault : capitolo.getAccertamento()) {
+          BigDecimal calculatedAmount;
+          if (ctAccertamentoDefault.getImporto().equals(BalanceDefaultAmountType.TOTAL.getType())) {
+            calculatedAmount = amountInstallment;
+          } else if (ctAccertamentoDefault.getImporto().contains(BalanceDefaultAmountType.EXTRACT_AMOUNT.getType())) {
+            engine.eval(ctAccertamentoDefault.getImporto());
+            Invocable invocable = (Invocable) engine;
+            String result = String.valueOf(invocable.invokeFunction(BalanceDefaultAmountType.EXTRACT_AMOUNT.getType(),
+              calculateAmountBalanceRequest.getRemittanceInformation()));
+            double resultDouble = Double.parseDouble(result);
+            calculatedAmount = BigDecimal.valueOf(resultDouble);
+          } else if (ctAccertamentoDefault.getImporto().contains(BalanceDefaultAmountType.CALCULATE_AMOUNT.getType())) {
+            engine.eval(ctAccertamentoDefault.getImporto());
+            Invocable invocable = (Invocable) engine;
+            String result = String.valueOf(
+              invocable.invokeFunction(BalanceDefaultAmountType.CALCULATE_AMOUNT.getType(), amountInstallment));
+            double resultDouble = Double.parseDouble(result);
+            calculatedAmount = BigDecimal.valueOf(resultDouble);
+          } else {
+            throw new InvalidValueException(ctAccertamentoDefault.getImporto() + " as function type to calculate amount balance not supported");
+          }
+          String amountString = Utilities.amountToString(calculatedAmount);
+          String amountWithoutSeparator = amountString.replace("\\.", "");
+          String amountWithSeparatorAndDecimal = amountWithoutSeparator.replace(",", ".");
+
+          ctAccertamentoDefault.setImporto(amountWithSeparatorAndDecimal);
+        }
+      }
+      return balanceDefaultMarshallingService.marshal(ctBilancioDefault);
```

Comment:
is the return of the function unmarshable as CtBilancio?
let's add this to service's test

Reply:
I mean, the XML String resolved should be a valid XML according the CtBilancio xsd...
So, inside the test, we should verify that the resolved template, which started with a CtDefaultBilancio object, could be unmarshalled as CtBilancio.

String result = service.calculateAmountBalance(request);

// Then
...
xmlUnmarshallerService.unmarshal(result, CtBilancio.class, jaxbContext, schema, NAMESPACE)
// this should not throw any exception, I used the service (not the mock) on this example, you could use the real service or directly the jaxb code